cis194/week6/zehua/src/Exercises.hs

```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# OPTIONS_GHC -fno-warn-missing-methods #-}
+module Exercises where
+
+
+-- ex1
+fib :: Integer -> Integer
+fib 0 = 0
+fib 1 = 1
+fib n = fib (n - 1) + fib (n - 2)
+-- fib = (+) <$> fib . subtract 1 <*> fib . subtract 2
+
+fibs1 :: [Integer]
+fibs1 = map fib [0..]
+
+
+-- ex2
+fibs2 :: [Integer]
+fibs2 = map fst $ iterate fibnext (0,1)
+  where
+    fibnext (n1, n) = (n, n1 + n)
+
+
+-- ex3
+data Stream a = Cons a (Stream a) deriving Eq
+
+streamToList :: Stream a -> [a]
+streamToList (Cons a b) = a : streamToList b
+
+instance Show a => Show (Stream a) where
+  show = show . take 20 . streamToList
+
+
+-- ex4
+streamRepeat :: a -> Stream a
+streamRepeat a = Cons a $ streamRepeat a
+
+streamMap :: (a -> b) -> Stream a -> Stream b
+streamMap f (Cons a b) = Cons (f a) $ streamMap f b
+
+streamFromSeed :: (a -> a) -> a -> Stream a
+streamFromSeed f a = Cons a $ streamFromSeed f (f a)
+
+
+-- ex5
+nats :: Stream Integer
+nats = streamFromSeed (+1) 1
+
+interleaveStreams :: Stream a -> Stream a -> Stream a
+interleaveStreams (Cons a b) s = (Cons a (interleaveStreams s b))
+-- this version is not as lazy and actually does not work for ruler
+-- interleaveStreams (Cons x1 y1) (Cons x2 y2) = (Cons x1 (Cons x2 (interleaveStreams y1 y2)))
+
+ruler :: Stream Integer
+-- interleaving with all 0 and itself but plus 1 to all elements
+ruler = interleaveStreams (streamRepeat 0) (streamMap (+1) ruler)
+
+-- ex6
+-- x = 0 + 1x + 0x^2 + 0x^3 + ...
+x :: Stream Integer
+x = Cons 0 $ Cons 1 $ streamRepeat 0
+
+instance Num (Stream Integer) where
+  fromInteger n                 = Cons n (streamRepeat 0)
+  negate (Cons n xs)            = Cons (-n) xs
+  (Cons a0 a') + (Cons b0 b')   = Cons (a0 + b0) $ a' + b'
+  (Cons a0 a') * b@(Cons b0 b') = Cons (a0 * b0) $ (fromInteger a0) * b' + a' * b
+
+--  (a0/b0) + x((1/b0)(A' − Q*B')).
+-- divStream :: Integral a => Stream a -> Stream a -> Stream a
```

Comment:
This one does not work. Giving the following error:

```
    • Couldn't match expected type ‘Integer’ with actual type ‘a’
      ‘a’ is a rigid type variable bound by
        the type signature for:
          divStream :: forall a.
                       Integral a =>
                       Stream a -> Stream a -> Stream a
        at src/Exercises.hs:70:1-59
    • In the first argument of ‘fromInteger’, namely ‘b0’
      In the second argument of ‘divStream’, namely ‘(fromInteger b0)’
      In the first argument of ‘(*)’, namely
        ‘(divStream (fromInteger 1) (fromInteger b0))’
    • Relevant bindings include
        q :: Stream a (bound at src/Exercises.hs:74:7)
        b' :: Stream a (bound at src/Exercises.hs:72:33)
        b0 :: a (bound at src/Exercises.hs:72:30)
        a' :: Stream a (bound at src/Exercises.hs:72:20)
        a0 :: a (bound at src/Exercises.hs:72:17)
        divStream :: Stream a -> Stream a -> Stream a
          (bound at src/Exercises.hs:72:1)
   |
75 |           (divStream (fromInteger 1) (fromInteger b0)) * (a' - q * b')
   |                                                   ^^
```

Looks like `fromInteger a` could not treat `Integral` as `Integer`

Reply:
`Integral` typeclass requires `Enum` and `Real` as constraints which `Stream a` does not have instance for.

Reply:
But the error msg complains about `a` instead of `Stream a`. `b0` is Integer` which is an instance of `Integral`.

Reply:
ok. I think the error message from ghc says exactly what is wrong. `fromInteger` expects exactly the data type `Integer` but we are passing in an `Integral a => a`, which `Integer` is an instance of but they are not the same thing.

Reply:
created #87 to make this work